core/codegen/src/derive/form_field.rs

```diff
@@ -322,7 +323,64 @@ pub fn validators<'v>(
             })).unwrap()
         });
 
-        Ok(exprs)
+    Ok(exprs)
+}
+
+pub fn default<'v>(
+    field: Field<'v>,
+) -> Result<Option<syn::Expr>> {
+    let mut exprs = FieldAttr::from_attrs(FieldAttr::NAME, &field.attrs)?
+        .into_iter()
+        .filter_map(|a| a.default)
+        .map(move |expr| {
+            // We need to define the default from the expression that the user has given. This means
+            // that the value that the user has written in the `#[field(default = ...)]` attribute
+            // must undergo the following logic:
+            // - If the value is the literal `None`, then the user want to unset the default given
+            //   by `FromFormField::default`. We must just return `None`.
+            // - If the value is a literal, we have to convert that literal into the type of the
+            //   field by calling `into` on it. This allows for better ergonomics by allowing the
+            //   user to write `#[form(default = "some value")]` for strings, instead of explicitly
+            //   having to call `to_owned`.
+            // - If the value is an integer literal, we have to add an intermediate step with an
+            //   extra type annotation. This is to make sure that a default of for example 4 for a
+            //   field of type `usize` gets interpreted as a `usize` instead of a `i32`. This step
+            //   must not be present for other types, as `"hello"` cannot be interpreted as a
+            //   `String`
+
+            use syn::{Expr, Lit, ExprLit};
+
+            let ty = &field.stripped_ty();
+            define_spanned_export!(expr.span => _form);
+
+            let is_int_lit = matches!(*expr, Expr::Lit(ExprLit { lit: Lit::Int(_), .. }));
+            let is_none = matches!(*expr, Expr::Path(ref e) if e.path.is_ident("None"));
+            if is_none {
+                syn::parse2(quote_spanned!(expr.span => {
+                    None
+                })).unwrap()
+            } else if is_int_lit {
+                syn::parse2(quote_spanned!(expr.span => {
+                    let __expr: Option<#ty> = (#expr).into();
+                    let __default: Option<#ty> = __expr.map(|v| v.into());
+                    __default
+                })).unwrap()
+            } else {
+                syn::parse2(quote_spanned!(expr.span => {
+                    let __default: Option<#ty> = Some({ #expr }.into());
+                    __default
+                })).unwrap()
+            }
+        });
```

Comment:
```suggestion
            use syn::{Expr, Lit, ExprLit};

            // As a result of calling `#expr.into()`, type inference fails for
            // two common expressions: integer literals and the bare `None`. As
            // a result, we cheat: if the syntax matches either of these two
            // conditions, we provide the field type as a hint.
            let is_int_lit = matches!(*expr, Expr::Lit(ExprLit { lit: Lit::Int(_), .. }));
            let is_none = matches!(*expr, Expr::Path(ref e) if e.path.is_ident("None"));
            let ty = field.stripped_ty();
            let ty_hint = (is_int_lit || is_none)
                .then(|| quote!(#ty))
                .unwrap_or_else(|| quote!(_));

            syn::parse2(quote_spanned!(expr.span => {
                let __expr: Option<#ty_hint> = #expr.into();
                let __default: Option<#ty> = __expr.map(|v| v.into());
                __default
            })).unwrap()
        });
```

Reply:
This change causes some type inference edge cases to fail again:

```rust
#[derive(rocket::FromForm)]
struct Form {
    #[field(default = 10 + 2)]
    field2: i8,
}
```
Will currently expand into:
```rust
.unwrap_or_else(|| {
    {
        let __default: Option<i8> = Some({ 10 + 2 }.into());
        __default
    }
    .or_else(|| <i8 as ::rocket::form::prelude::FromForm<'__f>>::default(_opts))
    .ok_or_else(|| ::rocket::form::prelude::ErrorKind::Missing.into())
})
```
With this change it expands into:
```rust
.unwrap_or_else(|| {
    {
        let __expr: Option<_> = 10 + 2.into();
        let __default: Option<i8> = __expr.map(|v| v.into());
        __default
    }
    .or_else(|| <i8 as ::rocket::form::prelude::FromForm<'__f>>::default(_opts))
    .ok_or_else(|| ::rocket::form::prelude::ErrorKind::Missing.into())
})
```
which fails to compile with:
```
error[E0282]: type annotations needed
   --> src/main.rs:122:49
    |
122 |         let __default: Option<i8> = __expr.map(|v| v.into());
    |                                                 ^ consider giving this closure parameter a type
    |
    = note: type must be known at this point
```

Reply:
Hmm second issue:
```rust
#[derive(rocket::FromForm)]
struct Form {
    #[field(default = Some(true))]
    field2: Option<bool>,
}
```
Used to get expanded to:
```rust
.unwrap_or_else(|| {
    {
        let __default: Option<Option<bool>> = Some({ Some(true) }.into());
        __default
    }
    .or_else(|| {
        <Option<bool> as ::rocket::form::prelude::FromForm<'__f>>::default(
            _opts,
        )
    })
    .ok_or_else(|| ::rocket::form::prelude::ErrorKind::Missing.into())
})
```
whereas now it expands to:
```rust
.unwrap_or_else(|| {
    {
        let __expr: Option<_> = { Some(true) }.into();
        let __default: Option<Option<bool>> = __expr.map(|v| v.into());
        __default
    }
    .or_else(|| {
        <Option<bool> as ::rocket::form::prelude::FromForm<'__f>>::default(
            _opts,
        )
    })
    .ok_or_else(|| ::rocket::form::prelude::ErrorKind::Missing.into())
})
```
This fails to compile with:
```
error[E0282]: type annotations needed
   --> src/main.rs:116:59
    |
116 |         let __default: Option<Option<bool>> = __expr.map(|v| v.into());
    |                                                           ^ consider giving this closure parameter a type
    |
    = note: type must be known at this point
```
I am not sure how you'd like to see this resolved though.